### ISSUE TITLE
unused bindings

### ISSUE THREAD

**Issue by borkdude** [borkdude/clj-kondo#173]
Report unused bindings

TODO:

- [x] suppress warning when using a binding that starts with an underscore, e.g. `(fn [_] "foo")`
- [x] false positive: `(fn [{:keys [x] :or {x 1}}] x)`
- [x] false positive: `(let [{:keys [patient/id order/id]} {}] id)` (both are unused, while in reality only one is unused)
- [x] support `loop`, `if-let`, `when-let`
- [x] support for `for`, `doseq`, `with-open`
- [x] handle `:let`, `:when` and `:while` in `for` and `doseq`
- [x] look more closely to defmacro locals + syntax-quote
- [x] emit warnings when something is not a valid binding form
- [x] fix false positives with locals used in metadata

Linux binary: https://3192-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.27-alpha-SNAPSHOT-linux-amd64.zip
Mac binary: https://3194-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.27-alpha-SNAPSHOT-macos-amd64.zip

**Comment by borkdude** [comment]
@mynomoto A very preliminary version of unused bindings is supported in this linux binary: https://3064-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.27-alpha-SNAPSHOT-linux-amd64.zip
Since you asked for this feature, I'd appreciate it if you could test a bit with it and report your findings. As of now it only reports unused bindings in `let`, `fn` and `defn`.

**Comment by mynomoto** [comment]
Sure, will do and report.

**Comment by borkdude** [comment]
Binary link updated with fixes.

**Comment by mynomoto** [comment]
Found one false positive
```
(defn false-positive-bind
  [{:keys [:x]}]
  (if-let [x (identity x)]
    x
    :y))
```
Reports: `cljkondo.clj:2:12: warning: unused binding x`

**Comment by mynomoto** [comment]
Changing `if-let` to `when-let` causes the same problem. It's not related to the destructuring of `:x`.

**Comment by borkdude** [comment]
@mynomoto Please ensure you're using the latest binary linked in the top comment. I believe I've already fixed this in the last few hours :-).

**Comment by mynomoto** [comment]
I just finished linting a project from work and no more false positives after the upgrade. :shipit:  Thanks!

**Comment by mynomoto** [comment]
I would love for this to cover cats.core/mlet (it works just like a let) but I'm not sure if you intend to or not in this first version. I wonder if that could be configurable for well behaved forms.

**Comment by borkdude** [comment]
@mynomoto That should work when using `:lint-as {cats.core/mlet clojure.core/let}` in the configuration.

**Comment by borkdude** [comment]
Fixed with 6d9020a083ec2cc9e66ddbff4d06b1e88d4a1d34

**Comment by mynomoto** [comment]
> @mynomoto That should work when using :lint-as {cats.core/mlet clojure.core/let} in the configuration.
Cool, I will try that! Thanks!

**Comment by mynomoto** [comment]
Found a false positive related to metadata:
```
(defn false-positive-metadata
  [a b]
  ^{:key (str a b)} [:other])
```
cljkondo.clj:2:4: warning: unused binding a
cljkondo.clj:2:6: warning: unused binding b

**Comment by borkdude** [comment]
Good catch! I'll look into it.

**Comment by borkdude** [comment]
@mynomoto Fixed in https://3192-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.27-alpha-SNAPSHOT-linux-amd64.zip
Can you test? I also found this catched a false positive here:
https://github.com/clojure/clojure/blob/8c402a8c9695a4eddc07cbbe0d95d44e1372f0bf/src/clj/clojure/zip.clj#L31

**Comment by mynomoto** [comment]
The false positive fix worked, as the `:lint-as` configuration. Thanks!

**Comment by svdo** [comment]
I think I'm getting a false positive on my CLJS code. I'm using a default import (`:require ["react-native-view-overflow" :default ViewOverflow]`) and this gives a warning: "namespace react-native-view-overflow is required but never used".

Other than that no surprises :)
